examples/virtualmachines/main.go

```diff
@@ -0,0 +1,180 @@
+be used to navigate a vSphere inventory structure using govmomi.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25/mo"
+)
+
+// GetEnvString returns string from environment variable.
+func GetEnvString(v string, def string) string {
+	r := os.Getenv(v)
+	if r == "" {
+		return def
+	}
+
+	return r
+}
+
+// GetEnvBool returns boolean from environment variable.
+func GetEnvBool(v string, def bool) bool {
+	r := os.Getenv(v)
+	if r == "" {
+		return def
+	}
+
+	switch strings.ToLower(r[0:1]) {
+	case "t", "y", "1":
+		return true
+	}
+
+	return false
+}
+
+const (
+	envURL      = "GOVMOMI_URL"
+	envUserName = "GOVMOMI_USERNAME"
+	envPassword = "GOVMOMI_PASSWORD"
+	envInsecure = "GOVMOMI_INSECURE"
+)
+
+var urlDescription = fmt.Sprintf("ESX or vCenter URL [%s]", envURL)
+var urlFlag = flag.String("url", GetEnvString(envURL, "https://username:password@host/sdk"), urlDescription)
+
+var insecureDescription = fmt.Sprintf("Don't verify the server's certificate chain [%s]", envInsecure)
+var insecureFlag = flag.Bool("insecure", GetEnvBool(envInsecure, false), insecureDescription)
+
+func processOverride(u *url.URL) {
+	envUsername := os.Getenv(envUserName)
+	envPassword := os.Getenv(envPassword)
+
+	// Override username if provided
+	if envUsername != "" {
+		var password string
+		var ok bool
+
+		if u.User != nil {
+			password, ok = u.User.Password()
+		}
+
+		if ok {
+			u.User = url.UserPassword(envUsername, password)
+		} else {
+			u.User = url.User(envUsername)
+		}
+	}
+
+	// Override password if provided
+	if envPassword != "" {
+		var username string
+
+		if u.User != nil {
+			username = u.User.Username()
+		}
+
+		u.User = url.UserPassword(username, envPassword)
+	}
+}
+
+func exit(err error) {
+	fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	os.Exit(1)
+}
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	flag.Parse()
+
+	// Parse URL from string
+	u, err := url.Parse(*urlFlag)
+	if err != nil {
+		exit(err)
+	}
+
+	// Override username and/or password as required
+	processOverride(u)
+
+	// Connect and log in to ESX or vCenter
+	c, err := govmomi.NewClient(ctx, u, *insecureFlag)
+	if err != nil {
+		exit(err)
+	}
+
+	// to know when its finished
+	var wg sync.WaitGroup
+
+	ch := make(chan string)
+
+	// Get view of VirtualMachine
+	m := view.NewManager(c.Client)
+
+	v, err := m.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"VirtualMachine"}, true)
+	if err != nil {
+		exit(err)
+	}
+
+	// Get summary property (VirtualMachineSummary)
+	// Reference for this data object:
+	// https://www.vmware.com/support/developer/vc-sdk/visdk25pubs/ReferenceGuide/vim.vm.Summary.html
+
+	var vms []mo.VirtualMachine
+	err = v.Retrieve(ctx, []string{"VirtualMachine"}, []string{"summary"}, &vms)
+	if err != nil {
+		exit(err)
+	}
+
```

Comment:
At this point, you have all the data in memory, so there's no need for the go routines.  Just the for loop + print is fine.